// MODULE GENE READ NVC RSEQC
READNVC = """
    export PYTHONPATH=/usr/local/cluster/python/usr/lib64/python2.6/site-packages/:\$PYTHONPATH && 
    /usr/local/cluster/python2.7/bin/python2.7 /usr/local/cluster/python/usr/bin/read_NVC.py
""".stripIndent().trim()

@preserve
rseqc_read_NVC_gfu = 
{
    doc title: "GFU: rseqc quality control of bam files: read_quality",
        desc: """
            This module is used to check the nucleotide composition bias. 
            Due to random priming, certain patterns are over represented at the beginning (5â€™end) of reads. 
            This bias could be easily examined by NVC (Nucleotide versus cycle) plot. 
            NVC plot is generated by overlaying all reads together, then calculating nucleotide 
            composition for each position of read (or each sequencing cycle). 
            In ideal condition (genome is random and RNA-seq reads is randomly sampled from genome), 
            we expect A%=C%=G%=T%=25% at each position of reads.
        """,
        constrains: "I am forcing export of site-packages to get qcmodule.",
        author: "davide.rambaldi@gmail.com"

    produce("*.NVC_plot.pdf","*.NVC_plot.r","*.NVC.xls") {
        exec"""
            echo -e "[rseqc_read_NVC]: NVC bias on $input.bam";
            $READNVC -i $input.bam -o $input.prefix;
        """
    }
    forward input.bam
}